### PR TITLE
feat: templator view file finder

### PR DESCRIPTION
**Pull request by SonyPradana**

### CODE REVIEW

File: src/System/View/TemplatorFinder.php [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+     */
+    public function exists(string $view_name): bool
+    {
+        if (isset($this->views[$view_name])) {
+            return true;
+        }
+
+        foreach ($this->paths as $path) {
+            foreach ($this->extensions as $extenstion) {
+                if (file_exists($path . $view_name . $extenstion)) {
```
Comment (SonyPradana) [inline src/System/View/TemplatorFinder.php:69]: miss delimater, and batter add realpath.

```php
    /**
     * Resolve the path.
     *
     * @param  string  $path
     * @return string
     */
    protected function resolvePath($path)
    {
        return realpath($path) ?: $path;
    }
```

File: src/System/View/TemplatorFinder.php [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+            $this->paths[] = $path;
+        }
+
+        return $this;
+    }
+
+    /**
+     * Add exteention in first array.
+     */
+    public function addExctension(string $extension): self
```
Comment (SonyPradana) [inline src/System/View/TemplatorFinder.php:113]: Typo

File: src/System/View/AbstractTemplatorParse.php [hunk 3]
```diff
@@ -6,18 +6,19 @@
...
+
     /**
      * Uses poller.
      *
      * @var string[]
      */
     protected $uses = [];
 
-    final public function __construct(string $templateDir, string $cacheDir)
+    final public function __construct(TemplatorFinder $finder, string $cacheDir)
```
Comment (anggerpradana) [inline src/System/View/AbstractTemplatorParse.php:19]: Backwards compatibility please add string parameter support.

File: src/System/View/Templator.php [hunk 4]
```diff
@@ -19,14 +18,14 @@
...
 class Templator
 {
-    private string $templateDir;
+    protected TemplatorFinder $finder;
     private string $cacheDir;
     public string $suffix = '';
     public int $max_depth = 5;
 
-    public function __construct(string $templateDir, string $cacheDir)
+    public function __construct(TemplatorFinder $finder, string $cacheDir)
```
Comment (anggerpradana) [inline src/System/View/Templator.php:26]: Backwards compatibility please add string parameter support.

File: src/System/View/TemplatorFinder.php [hunk 5]
```diff
@@ -0,0 +1,143 @@
+<?php
```
Comment (anggerpradana) [inline src/System/View/TemplatorFinder.php:1]: better with

- `get` file get content.
- `getPaths` and `getEctension`

**Review by anggerpradana — CHANGES_REQUESTED** [review]
so huge changes please provide backwash compatibility.

File: src/System/View/TemplatorFinder.php (hunk 5 above)
Comment (SonyPradana) [reply src/System/View/TemplatorFinder.php:1]: I don't like add get, to get file content. lets Templator do this.

File: src/System/View/Templator.php (hunk 4 above)
Comment (SonyPradana) [reply src/System/View/Templator.php:26]: nice

File: src/System/View/AbstractTemplatorParse.php (hunk 3 above)
Comment (SonyPradana) [reply src/System/View/AbstractTemplatorParse.php:19]: we don't need this because is an internal.

File: src/System/View/TemplatorFinder.php [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+     *
+     * @param string[] $paths
+     * @param string[] $extensions
+     */
+    public function __construct(array $paths, ?array $extensions = null)
+    {
+        foreach ($paths as $path) {
+            $this->paths[] = $this->resolvePath($path);
+        }
+        $this->extensions = $extensions ?? ['.temlate.php'];
```
Comment (anggerpradana) [inline src/System/View/TemplatorFinder.php:43]: another typo

File: src/System/View/TemplatorFinder.php [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+     * Add path to posible path location.
+     */
+    public function addPath(string $path): self
+    {
+        if (!isset($this->paths[$path])) {
+            $this->paths[] = $this->resolvePath($path);
+        }
+
+        return $this;
+    }
```
Comment (anggerpradana) [inline src/System/View/TemplatorFinder.php:112]: add more test

File: src/System/View/Templator.php [hunk 8]
```diff
@@ -19,15 +18,21 @@
...
+     * Create new intance.
+     *
+     * @param TemplatorFinder|string $finder if String will genarte TemplatorFinder with empty extension
+     */
+    public function __construct($finder, string $cacheDir)
     {
-        $this->templateDir = $templateDir;
-        $this->cacheDir    = $cacheDir;
+        // Backwards compatibility with templator finder.
+        $this->finder    = is_string($finder) ? new TemplatorFinder([$finder], ['']) : $finder;
```
Comment (anggerpradana) [inline src/System/View/Templator.php:34]: use default extension give null.

**Review by anggerpradana — CHANGES_REQUESTED** [review]

**Review by anggerpradana — APPROVED** [review]

**Pull request merged**